### ISSUE TITLE
Boolean flag in Javadoc Checks that shows a current Javadoc comment has unclosed HTML tags

### ISSUE THREAD

**Issue by baratali** [checkstyle/checkstyle#3311]
Add a boolean flag (hasUnclosedTag) to Javadoc Checks that shows a current Javadoc comment has unclosed HTML tags (and therefore has non-nested AST structure).

Reason:

> We expect non XHTML structure(we are not failing) so all Javadoc check have to expect non-nested AST. It will create a lot of problems to write reliable Javadoc Check, as amount of false positives will be huge.
> We need somehow to allow Check to decide to skip execution on Javadoc that is not XHTML , to let author of Check decide to support non-nested AST or not. 
> Can we somehow grab from parser a flag that javadoc is not xhtml ? 

Probably that flag should be obtained in the JavadocParser grammar. Have a look at this rule:

``` antlr
htmlElement: htmlTag
            | singletonElement
            | paragraph
            | li
            | tr
            | td
            | th
            | body
            | colgroup
            | dd
            | dt
            | head
            | html
            | option
            | tbody
            | thead
            | tfoot

            | pTagOpen ### Here goes unclosed tags ###
            | liTagOpen
            | trTagOpen
            | tdTagOpen
            | thTagOpen
            | bodyTagOpen
            | colgroupTagOpen
            | ddTagOpen
            | dtTagOpen
            | headTagOpen
            | htmlTagOpen
            | optionTagOpen
            | tbodyTagOpen
            | theadTagOpen
            | tfootTagOpen

            | pTagClose
            | liTagClose
            | trTagClose
            | tdTagClose
            | thTagClose
            | bodyTagClose
            | colgroupTagClose
            | ddTagClose
            | dtTagClose
            | headTagClose
            | htmlTagClose
            | optionTagClose
            | tbodyTagClose
            | theadTagClose
            | tfootTagClose
            ;
```
## 

Result changes to [AbstractJavadocCheck.java](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java):
filed - `violateExecutionOnNonTightHtml`
method - `acceptJavadocWithNonTightHtml()`

**Comment by unrealwork** [comment]
I am on it!

**Comment by ps-sp** [comment]
@rnveach @romani @Vladlis It looks like this issue can be resolved both ways, by adding code to grammar and by post processing the Parse Tree. Do you think that adding java code to grammar would be a bad desgin which may later create issues while upgrading ANTLR grammar for javadoc as it is creating issues for java right now ?

**Comment by rnveach** [comment]
Just my opinion, but if it makes things faster having Java code inside the grammar, i am fine with it.
We should limit Java code in grammar since grammar is not suppose to rely on external language, but it is our grammar so it shouldn't be a big deal.

@romani why do we need this boolean for unclosed tags? Checking the parent and position will tell anyone if the tag is closed.
Examples:
````
    |  |  |  |      |--HTML_ELEMENT -> <p>Test</p> [2:4]
    |  |  |  |      |  `--PARAGRAPH -> <p>Test</p> [2:4]
    |  |  |  |      |      |--P_TAG_OPEN -> <p> [2:4]
    |  |  |  |      |      |  |--OPEN -> < [2:4]
    |  |  |  |      |      |  |--P_HTML_TAG_NAME -> p [2:5]
    |  |  |  |      |      |  `--CLOSE -> > [2:6]
    |  |  |  |      |      |--TEXT -> Test [2:7]
    |  |  |  |      |      `--P_TAG_CLOSE -> </p> [2:11]
    |  |  |  |      |          |--OPEN -> < [2:11]
    |  |  |  |      |          |--SLASH -> / [2:12]
    |  |  |  |      |          |--P_HTML_TAG_NAME -> p [2:13]
    |  |  |  |      |          `--CLOSE -> > [2:14]
    |  |  |  |      |--HTML_ELEMENT -> <p> [2:15]
    |  |  |  |      |  `--P_TAG_OPEN -> <p> [2:15]
    |  |  |  |      |      |--OPEN -> < [2:15]
    |  |  |  |      |      |--P_HTML_TAG_NAME -> p [2:16]
    |  |  |  |      |      `--CLOSE -> > [2:17]
````
Could this be a lazy populated field instead like [`DetailAST.getBranchTokenTypes`](https://github.com/checkstyle/checkstyle/blob/4a087c020c2b97130e94f9292b94cb4bc1f6ed93/src/main/java/com/puppycrawl/tools/checkstyle/api/DetailAST.java#L342)

**Comment by ps-sp** [comment]
@rnveach Please have a look at [#4390 comment](https://github.com/checkstyle/checkstyle/issues/4390#issuecomment-310900590)

**Comment by Vladlis** [comment]
I've tried a few times, but could not understand the problem raised in this issue. And what exactly new flag would mean? That have `<foo>` without `</foo>`?
@romani @baratali please explain with examples

**Comment by baratali** [comment]
@Vladlis 
Unclosed HTML tag can change Parse tree's structure. If a tag is an unclosed, its child content won't be parsed as child nodes in the AST. It will be tricky to figure out what nodes are actually children, looking at AST. So, if there is a Check which has to verify, for example, presence of one specific tag inside another tag, it will probably need to skip Javadoc comments with unclosed tags, because it will be complicated to find out what nodes can be children of another nodes.

For example, how AST can change:
1. Closed paragraph (XHTML-style):
```
* <p>
* text <a>link</a>
* </p>
```
```
$ java -jar checkstyle-7.8.2-all.jar -j test.javadoc
JAVADOC -> * <p>\n* text <a>link</a>\n* </p>\n<EOF> [0:0]
|--LEADING_ASTERISK -> * [0:0]
|--TEXT ->   [0:1]
|--HTML_ELEMENT -> <p>\n* text <a>link</a>\n* </p> [0:2]
|   `--PARAGRAPH -> <p>\n* text <a>link</a>\n* </p> [0:2]
|       |--P_TAG_OPEN -> <p> [0:2]
|       |   |--OPEN -> < [0:2]
|       |   |--P_HTML_TAG_NAME -> p [0:3]
|       |   `--CLOSE -> > [0:4]
|       |--NEWLINE -> \n [0:5]
|       |--LEADING_ASTERISK -> * [1:0]
|       |--TEXT ->  text  [1:1]
|       |--HTML_TAG -> <a>link</a> [1:7]
|       |   |--HTML_ELEMENT_OPEN -> <a> [1:7]
|       |   |   |--OPEN -> < [1:7]
|       |   |   |--HTML_TAG_NAME -> a [1:8]
|       |   |   `--CLOSE -> > [1:9]
|       |   |--TEXT -> link [1:10]
|       |   `--HTML_ELEMENT_CLOSE -> </a> [1:14]
|       |       |--OPEN -> < [1:14]
|       |       |--SLASH -> / [1:15]
|       |       |--HTML_TAG_NAME -> a [1:16]
|       |       `--CLOSE -> > [1:17]
|       |--NEWLINE -> \n [1:18]
|       |--LEADING_ASTERISK -> * [2:0]
|       |--TEXT ->   [2:1]
|       `--P_TAG_CLOSE -> </p> [2:2]
|           |--OPEN -> < [2:2]
|           |--SLASH -> / [2:3]
|           |--P_HTML_TAG_NAME -> p [2:4]
|           `--CLOSE -> > [2:5]
|--NEWLINE -> \n [2:6]
`--EOF -> <EOF> [3:0]
```
2. Unclosed paragraph:
```
* <p>
* text <a>link</a>
```
```
$ java -jar checkstyle-7.8.2-all.jar -j test.javadoc
JAVADOC -> * <p>\n* text <a>link</a>\n<EOF> [0:0]
|--LEADING_ASTERISK -> * [0:0]
|--TEXT ->   [0:1]
|--HTML_ELEMENT -> <p> [0:2]
|   `--P_TAG_OPEN -> <p> [0:2]
|       |--OPEN -> < [0:2]
|       |--P_HTML_TAG_NAME -> p [0:3]
|       `--CLOSE -> > [0:4]
|--NEWLINE -> \n [0:5]
|--LEADING_ASTERISK -> * [1:0]
|--TEXT ->  text  [1:1]
|--HTML_ELEMENT -> <a>link</a> [1:7]
|   `--HTML_TAG -> <a>link</a> [1:7]
|       |--HTML_ELEMENT_OPEN -> <a> [1:7]
|       |   |--OPEN -> < [1:7]
|       |   |--HTML_TAG_NAME -> a [1:8]
|       |   `--CLOSE -> > [1:9]
|       |--TEXT -> link [1:10]
|       `--HTML_ELEMENT_CLOSE -> </a> [1:14]
|           |--OPEN -> < [1:14]
|           |--SLASH -> / [1:15]
|           |--HTML_TAG_NAME -> a [1:16]
|           `--CLOSE -> > [1:17]
|--NEWLINE -> \n [1:18]
`--EOF -> <EOF> [2:0]
```
``text`` and ``<a>`` are still under paragraph, however, they are parsed as sibling nodes, not children nodes to ``<p>``.

**Comment by Vladlis** [comment]
@baratali thanks, now I see the meaning of this statement
> to let author of Check decide to support non-nested AST or not.

@ps-sp 
Performance is in priority. But supposing there is no difference in what way to solve this issue, I'd prefer modifying post-processing rather than grammar to keep the latter clearer.

**Comment by ps-sp** [comment]
@baratali
Do you know if there is specific reason for which inside all the nested elements [`htmlElement rule`](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4#L54-#L103) hasn't been utilized and _htmlElements_ have been listed separately like this - [JavadocParser.g4 L137-L166](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4#L137-#L166). Also, [`...TagClose rules`](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4#L88-#L103) haven't been included in nested elements, but [`...TagOpen rules`](https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4#L153-#L166) have been, why so (certain HTML elements, like `<tbody>` and `<colgroup>` can rightfully have their start tag [omitted](https://www.w3.org/TR/html51/syntax.html#optional-start-and-end-tags) if certain conditions are met) ?

**Comment by romani** [comment]
@ps-sp , 

please remove chapter - http://checkstyle.sourceforge.net/writingjavadocchecks.html#Boolean_flag_indicating_the_presence_of_unclosed_HTML_tags
from web documentation.

**Comment by ps-sp** [comment]
@romani 
ok. will do.

**Comment by pbludov** [comment]
https://github.com/checkstyle/checkstyle/issues/1112 is still broken. Tested with config
```
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="SummaryJavadoc">
      <property name="violateExecutionOnNonTightHtml" value="true" />
    </module>
  </module>
</module>
```

**Comment by romani** [comment]
@ps-sp , please provide update to documentation , to let us close this issue.

--------------

@pbludov ,

> is still broken. Tested with config

please open new issue.

**Comment by pbludov** [comment]
>please open new issue.
@romani There is already one. The https://github.com/checkstyle/checkstyle/issues/1112 itself.

**Comment by romani** [comment]
@ps-sp , please update documentation, all related checks at http://checkstyle.sourceforge.net/config_javadoc.html should have new properties.

**Comment by ps-sp** [comment]
@romani 
>all related checks at http://checkstyle.sourceforge.net/config_javadoc.html should have new properties

All related checks already seem to be having the property `violateExecutionOnNonTightHtml`.

**Comment by romani** [comment]
Confirmed, thanks a lot.
Not sure what Check forced me to write such message, might be some old one, my bad.

**Comment by romani** [comment]
documentation is added to 8.8 version.

**Comment by rnveach** [comment]
Related checks are currently: AtclauseOrder, JavadocParagraph, JavadocTagContinuationIndentation, NonEmptyAtclauseDescription, SingleLineJavadoc, SummaryJavadoc
